[Concurrency] Updates after second SE pitch.

al45tair opened the pull request:
We're changing our approach to `Clock` support in executors such that a clock that isn't recognised by an executor will result in a call to an `enqueue` method on `Clock` itself; this will allow clocks to have control over exactly how they schedule jobs in cases where the executor doesn't already know what to do.

Also includes a small bug fix for `CooperativeExecutor` and some updated comments.

rdar://154191435

al45tair committed:
[Concurrency] Updates after second SE pitch.

We no longer attempt to convert timestamps from the passed-in `Clock`
in order to allow any clock to work with any executor.  Instead,
executors that do not recognise a clock should call the `enqueue`
function on that `Clock`, which lets the `Clock` itself decide how
to proceed.

Additionally, rename `SchedulableExecutor` to `SchedulingExecutor`.

al45tair committed:
[Concurrency] Add implementations of run and enqueue for built-in clocks.

The built-in clocks should have implementations of `run` and `enqueue`,
to allow derived clocks to call those implementations.

al45tair committed:
[Concurrency] Remove canonicalization code.

I don't think we actually need this.  If you have a non-canonical
(i.e. derived) clock, you can just implement `enqueue` and/or `run`
and call those methods on the clock you're wrapping.

al45tair committed:
[Concurrency] Fix CooperativeExecutor to not loop forever.

If a job enqueues another job on the executor, we might never leave
the inner `while` loop in the `run()` method.  Fix this by taking
the contents of the run queue and only running those jobs in the
queue at the time we enter the inner loop.

al45tair committed:
[Concurrency] Add a couple of comments.

Add some documentation comments to the Dispatch and CF executors,
and update the comments for the allocation and private data APIs.

al45tair commented:
@swift-ci Please smoke test

ktoso commented on stdlib/public/Concurrency/Clock.swift:
```diff
@@ -40,51 +40,61 @@ public protocol Clock<Duration>: Sendable {
   ///
-  /// - from duration: The `Duration` to convert
+  /// - job:         The job we wish to run
+  /// - at instant:  The time at which we would like it to run.
```
does spelling params like this actually parse well in docc? I thought one has to put only the second name 🤔

ktoso commented on stdlib/public/Concurrency/Clock.swift:
```diff
@@ -40,51 +40,61 @@ public protocol Clock<Duration>: Sendable {
+  /// given instant.
+  ///
+  /// The default implementation uses the `run` method to trigger a job that
+  /// does `executor.enqueue(job)`.  If a particular `Clock` knows that the
```
minor

```suggestion
  /// does `executor.enqueue(job)`.  If a particular ``Clock`` knows that the
```

maybe worth ref like that?

ktoso commented on stdlib/public/Concurrency/Clock.swift:
```diff
@@ -40,51 +40,61 @@ public protocol Clock<Duration>: Sendable {
+    run(trampoline, at: instant, tolerance: tolerance)
+  }
+
+  // Clocks that do not implement run will fatalError() if you try to use
```
```suggestion
  // Clocks that do not implement this method will fatalError() if you try to use
```

maybe? "implement run" sounds a bit weird mid-sentence

ktoso commented on test/Concurrency/Runtime/sleep_executor.swift:
```diff
@@ -37,8 +37,17 @@ final class TestExecutor: TaskExecutor, SchedulableExecutor, @unchecked Sendable
+
     // Convert to `Swift.Duration`
-    let duration = clock.convert(from: delay)!
+    let duration = delay as! Swift.Duration
```
can we put a comment here discussing the `as!`'s risk, could it ever fail?

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -441,6 +443,48 @@ extension ExecutorJob {
+// executor.
+extension ExecutorJob {
+
+  /// Create a trampoline to enqueue the specified job on the specified
```
minor?

```suggestion
  /// Create a trampoline to enqueue this job on the specified
```

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -441,6 +443,48 @@ extension ExecutorJob {
+  /// Create a trampoline to enqueue the specified job on the specified
+  /// executor.
+  ///
+  /// This is useful in conjunction with the `Clock.run()` API, which
```
I think this'll link well: 
```suggestion
  /// This is useful in conjunction with the ``Clock.run(_:at:tolerance:)`` API, which
```

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -441,6 +443,48 @@ extension ExecutorJob {
+  ///
+  /// Returns:
+  ///
+  /// A new ExecutorJob that will enqueue the specified job on the specified
```
```suggestion
  /// - A new ExecutorJob that will enqueue the specified job on the specified
```

needs the `-` I think?

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -441,6 +443,48 @@ extension ExecutorJob {
+  /// This is useful in conjunction with the `Clock.run()` API, which
+  /// runs a job on an unspecified executor.
+  ///
+  /// Parameters:
```
```suggestion
  /// - Parameters:
```

Needs the `-` I think?

ktoso commented on stdlib/public/Concurrency/DispatchGlobalExecutor.cpp:
```diff
@@ -300,7 +300,9 @@ platform_time(uint64_t nsec) {
   uint64_t deadline;
-  if (__builtin_mul_overflow(sec, NSEC_PER_SEC, &deadline)
+  if (sec < 0 || sec == 0 && nsec < 0)
+    deadline = 0;
```
👍

ktoso reviewed (approved):
lgtm, sorry about the docs nitpicks! I don't actually have a great way to preview those in tree I guess 🤔

al45tair commented:
> lgtm, sorry about the docs nitpicks! I don't actually have a great way to preview those in tree I guess 🤔

No worries. I think you're right about the doc comments; I'll go and update them all.

al45tair committed:
[SILGen][Concurrency] Improve DefaultExecutorFactory lookup.

Rather than just looking at top level in the module, start by searching
the type marked as `@main`.  This means that a library that provides a
protocol or superclass that the `@main` type can conform to can specify
an executor in a reasonable manner.

al45tair committed:
[Concurrency] Don't expose Dispatch or CF executors directly.

Also add `final` to the `CooperativeExecutor` declaration.

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair commented:
Linux looks like some WASI related problem, possibly to do with the cooperative executor?

al45tair committed:
[Tests][Concurrency] Remove clock canonicalization from test.

We've removed this code, so the test was failing.

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Fix cooperative executor to return only after all jobs run.

We were terminating after the first set of jobs; if one of them scheduled
another job, and there were no timers running, we would terminate,
which was wrong.

al45tair commented:
@swift-ci Please test

al45tair committed:
[Tests][Concurrency] Fix macOS build.

There was still some canonicalization in one of the tests, plus we needed
to update the ABI tests.

al45tair commented:
@swift-ci Please test

al45tair committed:
[Concurrency] Support continuous waits in CooperativeExecutor.

This fixes some WASM tests.

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci please test Windows platform

al45tair commented:
@swift-ci please test Windows platform

al45tair commented:
(Rebased)

al45tair committed:
[Concurrency] Fix minimal stdlib build again.

Some mentions of `ExecutorJob` were not guarded by the required
`SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY` condition.

al45tair commented:
@swift-ci Please test
